drivers/char/broadcom/vc_mem.c

```diff
@@ -216,132 +198,104 @@ static const struct file_operations vc_mem_fops = {
 };
 
 #ifdef CONFIG_DEBUG_FS
-static void vc_mem_debugfs_deinit(void)
-{
-	debugfs_remove_recursive(vc_mem_debugfs_entry);
-	vc_mem_debugfs_entry = NULL;
-}
-
-
-static int vc_mem_debugfs_init(
-	struct device *dev)
-{
-	vc_mem_debugfs_entry = debugfs_create_dir(DRIVER_NAME, NULL);
-	if (!vc_mem_debugfs_entry) {
-		dev_warn(dev, "could not create debugfs entry\n");
-		return -EFAULT;
-	}
-
-	debugfs_create_x32("vc_mem_phys_addr",
-				0444,
-				vc_mem_debugfs_entry,
-				(u32 *)&mm_vc_mem_phys_addr);
-	debugfs_create_x32("vc_mem_size",
-				0444,
-				vc_mem_debugfs_entry,
-				(u32 *)&mm_vc_mem_size);
-	debugfs_create_x32("vc_mem_base",
-				0444,
-				vc_mem_debugfs_entry,
-				(u32 *)&mm_vc_mem_base);
-
+static int vc_mem_ulong_get(void *data, u64 *val) {
+	*val = *(unsigned long*)data;
 	return 0;
 }
 
+DEFINE_DEBUGFS_ATTRIBUTE(vc_mem_phys_addr_fops, vc_mem_ulong_get, NULL, "0x%08llx\n");
 #endif /* CONFIG_DEBUG_FS */
 
-/* Module load/unload functions */
-
-static int __init
-vc_mem_init(void)
+static int vc_mem_probe(struct platform_device *pdev)
 {
-	int rc = -EFAULT;
-	struct device *dev;
-
-	pr_debug("%s: called\n", __func__);
-
-	mm_vc_mem_phys_addr = phys_addr;
-	mm_vc_mem_size = mem_size;
-	mm_vc_mem_base = mem_base;
-
-	pr_info("vc-mem: phys_addr:0x%08lx mem_base=0x%08x mem_size:0x%08x(%u MiB)\n",
-		mm_vc_mem_phys_addr, mm_vc_mem_base, mm_vc_mem_size,
-		mm_vc_mem_size / (1024 * 1024));
-
-	rc = alloc_chrdev_region(&vc_mem_devnum, 0, 1, DRIVER_NAME);
-	if (rc < 0) {
-		pr_err("%s: alloc_chrdev_region failed (rc=%d)\n",
-		       __func__, rc);
-		goto out_err;
+	int ret;
+	struct vc_mem *drv;
+	const __be32 *addrp;
+	int n_addr_bytes;
+
+	drv = devm_kzalloc(&pdev->dev, sizeof(*drv), GFP_KERNEL);
+	if (!drv)
+		return -ENOMEM;
+	platform_set_drvdata(pdev, drv);
+
+	// Note that the "reg" property provided by firmware is does not
+	// follow any conventions. It is just a bunch of 32-bit addresses.
+	addrp = of_get_property(pdev->dev.of_node, "reg", &n_addr_bytes);
+	if (!addrp || n_addr_bytes != 12)
+		return -ENODEV;
+	drv->base = be32_to_cpu(addrp[0]);
+	drv->size = be32_to_cpu(addrp[1]);
+	drv->phys_addr = be32_to_cpu(addrp[2]);
```

Comment:
Are you sure this is correct? On my Pi 4B `addrp[2] == 0xC0000000`. I think that's where the VC sees the start of SDRAM. I would also guess that the purpose served by `addrp[2]` is that on older Pi models it is an offset to add to addresses when doing DMA:
```
dma_memcpy(vc_addr, addrp[2] + host_addr, size);
```
However this is redundant because we have `dma-ranges`, which is the mechanism used by Linux to turn physical addresses into `dma_addr_t`.